README.md
# AsteroidsGame
Final project course of Functional Programming Utrecht University
# Asteroids

## Data and architecture design

By Matej Havranek & Teddy Gyabaah

## Introduction and Rules

#### The player controls a spaceship which can move around the screen. The spaceship has

#### to avoid the asteroids that pop up on the screen, and can destroy them by shooting at

#### them, at which point they split into smaller asteroids.

#### Each asteroid has a chance of having an enemy spaceship inside. When freed by

#### destroying the asteroid, it flies around and tries to shoot the player.

#### The goal of the player is to gain as much points as possible by destroying asteroids and

#### killing the enemies in the game.

## Design

### Data structures

#### The full state of the game is represented using a GameState object that contains three

#### fundamental states of our game, namely the state of the Game screen, the stats situation

#### and the actual objects involved in the game in a precise moment.

#### data​ ​GameState​ = ​State​ { ​gameScreen​ :: ​GameScreen​,


#### ​gameStats​ :: ​GameStats​,

#### ​gameObjects​ :: ​GameObjects​ }

#### data​ ​GameStats​ = ​Stats​ { ​name​ :: ​String​,

#### ​score​ :: ​Int​ }

#### data​ ​GameObjects​ = ​GameObjs​ { ​asteroids​ :: [​Asteroid​],

#### ​missiles​ :: [​Missile​],

#### ​enemies​ :: [​Player​],

#### ​player​ :: ​Player​ }

### The player

#### The player type in our Asteroids game will be represented by an enumeration, in this

#### definition we define that the player can be either controlled by a Human or by the

#### Computer.

#### data​ ​PlayerType​ = ​Human​ | ​Computer

#### The actual player is modeled by the Spaceship data type that contains all the properties

#### to handle it.

#### data​ ​Spaceship​ = ​Ship​ { ​shipPosition​ :: ​Point​,

#### ​shipSpeed​ :: ​Vector​,

#### ​angle​ :: ​Float​,

#### ​boundingBox​ :: ​BoundingBox​,

#### ​bulletTimeout​ :: ​Int​,

#### ​playerType​ :: ​PlayerType​ }

#### As we can see by design the enemies and the player differ only in the player type. The

#### player has to defeat enemies (by launching missiles at them) and destroy asteroids while


#### avoiding being hit by the asteroids or enemy missiles to get a higher score. Missiles are

#### defined as follows:

#### data​ ​Missile​ = ​Msl​ { ​lifetime​ :: ​Int​,

#### ​position​ :: ​Point​,

#### ​boundingBox​ :: ​BoundingBox​,

#### ​speed​ :: ​Vector​ }

#### And the function that launches them is

#### -- Create a missile heading in the direction of the spaceship

#### shoot​ :: ​Spaceship​ -> ​Missile

### Player movement

#### Asteroids is a game that has independently moving particles. This means that the

#### movement is continuous and can vary in direction and speed for every object on screen.

#### The player can decide where to move on the screen.

#### The move function takes the position of an object and a speed vector, and returns the

#### new position of the object.

#### -- Moves an object in the direction of the specified vector

#### move​ :: ​Point​ -> ​Vector​ -> ​Point

### Player adversary

#### As mentioned before the difference between the player and the enemies dwells in the

#### PlayerType. The enemies movements are controlled by the AI function that will decide

#### the speed and the direction of the ship in order to obtain a clear shot on the player.


#### Another fundamental adversary of the user are the Asteroids that float around the screen

#### modeled using the following structure:

#### data​ ​Asteroid​ = ​As​ { ​size​ :: ​Int​,

#### ​position​ :: ​Point​,

#### ​boundingBox​ :: ​BoundingBox​,

#### ​speed​ :: ​Vector​ }

#### An important property of Asteroids and Spaceships is the BoundingBox. The bounding

#### box is a rectangular border that fully encapsulates an object. Ii is important because it

#### allows us to detect collisions between the asteroids and the player by simply checking

#### whether their bounding boxes intersect.

#### -- Checks whether two bounding boxes collide

#### checkCollision​ :: ​BoundingBox​ -> ​BoundingBox​ -> ​Bool




game-cabal/ChangeLog.md
# Revision history for asteroids

## 0.1.0.0  -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


draft.hs
--------------------
------ Model -------
--------------------

--------------------
-- Defines the game screen we are currently on
--------------------
data GameScreen = MainScreen | PlayScreen | PauseScreen | EndScreen

--------------------
-- Defines whether an entity is controlled by a human or by AI (further AI types can be added)
--------------------
data PlayerType = Human | AI

--------------------
-- Defines the bounding box for an entity, simplifies collision detection
--------------------
data BoundingBox = Bound { leftUpper :: Point,
                           rightBottom :: Point }

--------------------
-- Defines a spaceship (player or enemy) and its properties
--------------------
data Spaceship = Ship { shipPosition :: Point,
                        shipSpeed :: Int,
                        angle :: Float,
                        shipBoundingBox :: BoundingBox,
                        bulletTimeout :: Int,
                        playerType :: PlayerType }

--------------------
-- Defines an asteroid and its properties
--------------------
data Asteroid = As { asSize :: Int,
                     asPosition :: Point,
                     asBoundingBox :: BoundingBox,
                     asSpeed :: Vector }

--------------------
-- Defines a missile (a single shot) and its properties
--------------------
data Missile = Msl { mslLifetime :: Int,
                     mslPosition :: Point, 
                     mslBoundingBox :: BoundingBox,
                     mslSpeed :: Vector }

--------------------
-- Contains the players name and score in the current game
--------------------
data GameStats = Stats { name :: String,
                         score :: Int }

--------------------
-- Encapsulates all the game objects (spaceships, asteroids, missiles) into one container
--------------------
data GameObjects = GameObjs { asteroids :: [Asteroid],
                              missiles :: [Missile],
                              enemies :: [Spaceship],
                              player :: Spaceship }

--------------------
-- Contains all information about the game and its state
--------------------
data GameState = State { gameScreen :: GameScreen,
                         gameStats :: GameStats,
                         gameObjects :: GameObjects }
--------------------
-- Returns the initial state of the game
--------------------
initialState :: GameState

--------------------
-- Creates a new ship (player/enemy) at the given point
--------------------
createShip :: PlayerType -> Point -> Spaceship

--------------------
-- Creates an asteroid of the given size at the given point
--------------------
createAsteroid :: Int -> Point -> Vector -> Asteroid

--------------------
-- Create a missile heading in the direction of the spaceship
--------------------
createMissile :: Spaceship -> Missile

--------------------
-- Adds a missile to the list of missiles in the gameObject
--------------------
addMissile :: GameObjects -> Missile -> GameObjects

--------------------
-- Sets the ships speed to propel it forward
--------------------
propelShip :: Spaceship -> Spaceship

--------------------
-- Moves an object in the direction of the specified vector
--------------------
move :: Point -> Vector -> Point

--------------------
-- Checks whether two bounding boxes collide
--------------------
checkCollision :: BoundingBox -> BoundingBox -> Bool

--------------------
-- Changes the spaceship angle and rotates the bounding box (graphic rotation is handled separately)
--------------------
rotateShip :: Spaceship -> Float -> Spaceship

--------------------
-- Whenever an asteroid is hit, explode it into smaller asteroids (or none if its too small)
--------------------
explodeAsteroid :: Asteroid -> IO [Asteroid]

--------------------
-- Updates the position of the asteroids
--------------------
updateAsteroids :: [Asteroid] -> [Asteroid]

--------------------
-- Updates the position of the missiles and remove them if their lifetimes are over
--------------------
updateMissiles :: [Missile] -> [Missile]


--------------------
------- View -------
--------------------

--------------------
-- Main rendering method. Based on game state determines the correct screen to render
--------------------
view :: GameState -> IO Picture

--------------------
-- Renders the first screen with game logo and player name input
--------------------
renderMainScreen :: GameState -> Picture

--------------------
-- Renders the game itself - player, enemies, asteroids, missiles and score
--------------------
renderPlayScreen :: GameState -> Picture

--------------------
-- Pause screen - shows the score and a message that the game is paused
--------------------
renderPauseScreen:: GameState -> Picture

--------------------
-- End screen - displays "game over", players name and final score
--------------------
renderEndScreen :: GameState -> Picture

--------------------
-- Converts current game stats (name and score) into a picture
--------------------
getStatsText :: GameStats -> Picture


--------------------
---- Controller ----
--------------------

--------------------
-- Update loop
--------------------
step :: Float -> GameState -> IO GameState

--------------------
-- Handle user input
--------------------
input :: Event -> GameState -> IO GameState

--------------------
-- Handle keypresses
--------------------
inputKey :: Event -> GameState -> GameState

--------------------
-- Handle mouse events (for the mouse input bonus)
--------------------
inputMouse :: Event -> GameState -> GameState

--------------------
-- AI determines the actions for enemy spaceships
--------------------
updateAI :: GameObjects -> GameObjects

--------------------
-- Performs collisions on all game entities
--------------------
performCollisions :: GameState -> GameState

game-cabal/src/Input.hs
module Input where

import Model

import qualified Data.Set as S
import Data.Char

import Graphics.Gloss
import Graphics.Gloss.Interface.IO.Game

--------------------
-- Handle user input
--------------------
input :: Event -> GameState -> IO GameState
input e gstate = return (inputKey e gstate)

--------------------
-- Handle keypresses
--------------------
inputKey :: Event -> GameState -> GameState
inputKey (EventKey (SpecialKey key) Down _ _) gstate = case key of

  -- ESC to exit game from pause screen or return to main screen from end screen
  KeyEsc -> case gameScreen gstate of
              PauseScreen -> gstate { gameScreen = EndScreen, gameIO = SaveHighscore }
              EndScreen -> resetToDefault gstate
              _ -> gstate

  -- Enter on main screen to start game
  KeyEnter -> case gameScreen gstate of
                MainScreen -> gstate { gameScreen = PlayScreen }
                EndScreen -> resetToDefault gstate
                _ -> gstate

  -- Space to shoot a missile
  KeySpace -> case gameScreen gstate of
                PlayScreen -> addPlayerMissile gstate
                _ -> gstate

  -- All other keys get added to keypress array
  _ -> gstate { gameKeys = S.insert key (gameKeys gstate) }

inputKey (EventKey (Char c) Down _ _) gstate = case toUpper c of
  -- P to pause/unpause the game
  'P' -> case gameScreen gstate of
          PlayScreen -> gstate { gameScreen = PauseScreen }
          PauseScreen -> gstate { gameScreen = PlayScreen }
          _ -> gstate

  'A' -> addRandomAsteroid gstate -- Adds random asteroids TODO: to remove just for debug
  'E' -> addRandomEnemies gstate -- Adds random enemies TODO: to remove just for debug

  _ -> gstate

-- Handle keys being released. Also specific events upon key release
inputKey (EventKey (SpecialKey key) Up _ _) gstate = case key of
  -- workaround to disable the flame here. TODO make this less ugly
  KeyUp -> case gameScreen gstate of
    PlayScreen -> gstate { gameObjects = (gameObjects gstate) { player = idleShip (player (gameObjects gstate)) }, gameKeys = S.delete key (gameKeys gstate) }
    _ -> gstate { gameKeys = S.delete key (gameKeys gstate) }

  _ -> gstate { gameKeys = S.delete key (gameKeys gstate) }

-- Handling mouse presses for left (propel ship) and right (shoot) buttons
inputKey (EventKey (MouseButton LeftButton) Down _ (_, _)) gstate = gstate { gameKeys = S.insert KeyUp (gameKeys gstate) } --simulate up arrow press
inputKey (EventKey (MouseButton LeftButton) Up _ (_, _)) gstate = gstate { gameObjects = (gameObjects gstate) { player = idleShip (player (gameObjects gstate)) }, gameKeys = S.delete KeyUp (gameKeys gstate) } --simulate up arrow release + flame workaround (TODO make prettier)

inputKey (EventKey (MouseButton RightButton) Down _ (_, _)) gstate = addPlayerMissile gstate --shoot

-- When mouse movement is detected, orient the player to face the mouse. TODO mouse offset is hardcoded, resolution changes are not allowed
inputKey (EventMotion (ptx, pty)) gstate = gstate { gameObjects = (gameObjects gstate) { player = turnToFace (player (gameObjects gstate)) (ptx + 640, pty + 360) } }

-- All other cases => do nothing
inputKey _ gstate = gstate

--------------------
-- Iterate over all keys that are pressed down and perform their respective actions
--------------------
handleAllKeys :: GameState -> GameState
handleAllKeys gstate = foldr handleKey gstate (gameKeys gstate)

--------------------
-- Handle individual keys and their respective actions
--------------------
handleKey :: SpecialKey -> GameState -> GameState
-- Up = propel ship
handleKey KeyUp gstate = case gameScreen gstate of
  PlayScreen -> gstate { gameObjects = (gameObjects gstate) { player = propelShip (player (gameObjects gstate)) } }
  _ -> gstate

-- Left = rotate left
handleKey KeyLeft gstate = case gameScreen gstate of
  PlayScreen -> gstate { gameObjects = (gameObjects gstate) { player = rotateShip (player (gameObjects gstate)) (-5) } }
  _ -> gstate

-- Right = rotate right
handleKey KeyRight gstate = case gameScreen gstate of
  PlayScreen -> gstate { gameObjects = (gameObjects gstate) { player = rotateShip (player (gameObjects gstate)) 5 } }
  _ -> gstate

-- default case for everything else
handleKey _ gstate = gstate

game-cabal/src/Score.hs
module Score where

import Model

import Control.Monad
import System.Directory

--------------------
-- Path to the highscore file
--------------------
highscoreFile :: String
highscoreFile = "highscore.txt"

--------------------
-- If the current score is higher than the saved highscore, rewrite it
--------------------
saveHighscore :: GameState -> IO ()
saveHighscore gstate = writeFile highscoreFile (show (gameHighscore gstate))

--------------------
-- Reads the current highscore stored in the highscore.txt file
--------------------
readHighscore :: IO Int
readHighscore = do
                  exists <- doesFileExist highscoreFile
                  Control.Monad.unless exists (writeFile highscoreFile "0")
                  txt <- readFile highscoreFile
                  return (read txt::Int)

game-cabal/src/Collisions.hs
module Collisions where

import Model

import Graphics.Gloss.Data.Point
import Graphics.Gloss.Data.Vector
import Graphics.Gloss.Geometry.Angle
import Graphics.Gloss.Interface.IO.Game
import qualified Graphics.Gloss.Data.Point.Arithmetic as GlossArithm

--------------------
-- Performs collisions on all game entities
--------------------
performCollisions :: GameState -> GameState
performCollisions gstate = collisionsAsteroidMissile (collisionsMissileEnemies (collisionsMissilePlayer (collisionsAsteroidPlayer gstate)))

--------------------
-- Resolves collisions between all missiles and all asteroids
--------------------
collisionsAsteroidMissile :: GameState -> GameState
collisionsAsteroidMissile gstate = incrementScore (gstate { gameObjects = newgo}) (length explosions)
    where
      go                    = gameObjects gstate
      ((newAsteroids,newMissile), explosions, newEnemies) = resolveCollisionAM gstate (asteroids go) (filter  (\m -> mslType m == PlayerMsl) (missiles go)) [] []
      newgo                 = go { asteroids = newAsteroids , missiles = newMissile ++ filter  (\m -> mslType m == EnemyMsl) (missiles go), enemies = newEnemies ++ enemies go ,animations= explosions++(animations go) }

      resolveCollisionAM :: GameState -> [Asteroid] -> [Missile] -> [Animation] -> [Enemy] -> (([Asteroid], [Missile]), [Animation], [Enemy])
      resolveCollisionAM  _ ax [] explod enms = ((ax, []), explod, enms)
      resolveCollisionAM  _ [] ms explod enms = (([], ms), explod, enms)
      resolveCollisionAM  gs (a:ax) (m:ms) explod enms | checkCollisionAM a m = ((explodedPieces ++ updasteroids, updmissiles), addedAnimation, addedEnemies)
                                                       | otherwise = ((a: updasteroids1, m:ms), explod, enms)
            where
              (explodedPieces,newRand) = explodeAsteroid gs a
              updatedGs = gs{randomGen = newRand}
              ((updasteroids,updmissiles),_,es) = resolveCollisionAM updatedGs ax ms addedAnimation es
              ((updasteroids1,updmissiles1),_,es1) = resolveCollisionAM updatedGs ax (m:ms) explod es1
              addedEnemies = if asSize a < 0.7 then enms else createEnemy (asPosition a) : enms
              addedAnimation = createAnimation(asPosition a):explod

--------------------
-- Resolves collisions between player and all asteroids
--------------------
collisionsAsteroidPlayer :: GameState -> GameState
collisionsAsteroidPlayer gstate | resolveCollisionPA ast pl = gameOver gstate
                                | otherwise = gstate
    where go = gameObjects gstate
          pl = player go
          ast = asteroids go

          resolveCollisionPA::[Asteroid] -> Spaceship -> Bool
          resolveCollisionPA xs pl = any (checkCollisionPA pl) xs

--------------------
-- Resolves collisions between all missiles and the player
--------------------

collisionsMissilePlayer :: GameState -> GameState
collisionsMissilePlayer gstate | resolveCollisionMP miss pl = gameOver gstate
                               | otherwise                             = gstate
    where go = gameObjects gstate
          pl = player go
          miss = missiles go

          resolveCollisionMP :: [Missile] -> Spaceship -> Bool
          resolveCollisionMP [] pl  = False
          resolveCollisionMP xs pl  = any (checkCollisionMP pl) xs

--------------------
-- Resolves collisions between all missiles and all enemies
--------------------
collisionsMissileEnemies :: GameState -> GameState
collisionsMissileEnemies gstate = incrementScore (gstate { gameObjects = newgo }) killed
    where
      go         = gameObjects gstate
      killed     = length (enemies go) - length newEnemies
      newEnemies = resolveCollisionMERec (enemies go) (filter (\m -> mslType m == PlayerMsl ) (missiles go))
      newgo      = go {enemies = newEnemies }

      resolveCollisionMERec :: [Enemy] -> [Missile] -> [Enemy]
      resolveCollisionMERec  sx []        = sx
      resolveCollisionMERec  [] _         = []
      resolveCollisionMERec (s:sx) (m:ms) | checkCollisionMP (enmShip s) m = resolveCollisionMERec sx ms
                                          | otherwise = s: resolveCollisionMERec sx (m:ms)

--------------------
-- Checks whether the player and an asteroid collide
--------------------
checkCollisionPA :: Spaceship -> Asteroid-> Bool
checkCollisionPA p a = magV (asPosition a GlossArithm.- shipPosition p) < (200 * asSize a)

--------------------
-- Checks whether the player and a missile collide
--------------------
checkCollisionMP :: Spaceship -> Missile-> Bool
checkCollisionMP p m = magV (mslPosition m GlossArithm.- shipPosition p) < 25

--------------------
-- Checks whether an asteroid and a missile collide
--------------------
checkCollisionAM :: Asteroid -> Missile-> Bool
checkCollisionAM a m = magV (asPosition a GlossArithm.- mslPosition m) < (280 * asSize a)


game-cabal/src/Controller.hs
-- | This module defines how the state changes
--   in response to time and user input
module Controller where

import Collisions
import Input
import Model
import Score

import Graphics.Gloss
import Graphics.Gloss.Interface.IO.Game

import qualified Data.Set as S
import Data.Char
import Data.Fixed

--------------------
-- Update loop - only update the game if the game is running, determined by the gameScreen we are in
--------------------
step :: Float -> GameState -> IO GameState
step secs gstate | gameScreen gstate == PlayScreen = handleIO (updateGame secs gstate { framesSinceLastSpawn = framesSinceLastSpawn gstate + 1} )
                 | otherwise = handleIO gstate

--------------------
-- Handles file operations (highscore load/save) based on flags in the gameState
--------------------
handleIO :: GameState -> IO GameState
handleIO gstate | gio == LoadHighscore = do
                                            hs <- readHighscore
                                            return gstate { gameHighscore = hs, gameIO = DoNothing }
                | gio == SaveHighscore = do
                                            saveHighscore gstate
                                            return gstate { gameIO = DoNothing }
                | otherwise = pure gstate

                where
                    gio = gameIO gstate

--------------------
-- Updates the AI, all game objects and then performs collision checking
--------------------
updateGame :: Float -> GameState -> GameState
updateGame secs gstate = spawnNewAsteroids secs $ performCollisions $ updateAI (gstate { gameObjects = updateGameObjects (gameObjects (handleAllKeys gstate)) })

--------------------
-- Periodically spawns new asteroids
--------------------
spawnNewAsteroids :: Float -> GameState -> GameState
spawnNewAsteroids secs gstate | framesSinceLastSpawn gstate > 100 = addRandomAsteroid gstate { framesSinceLastSpawn = 0 }
                              | otherwise = gstate

--------------------
-- Updates all the objects in the game (performs movement and roatation)
--------------------
updateGameObjects :: GameObjects -> GameObjects
updateGameObjects gobjs = gobjs { asteroids  = map move (asteroids gobjs),
                                  missiles   = updateMissiles (missiles gobjs),
                                  animations = updateAnimations (animations gobjs),
                                  enemies    = map move (enemies gobjs),
                                  player     = move (player gobjs) }


game-cabal/src/View.hs
{-# LANGUAGE InstanceSigs #-}
module View where

import Graphics.Gloss
import Model

--------------------
-- Main rendering method. Based on game state determines the correct screen to render
--------------------
view :: GameState -> IO Picture
view gstate = pure $ render gstate

--------------------
-- Type class for renderable objects
--------------------
class Renderable a where
  render :: a -> Picture

-- Renders the game itself
instance Renderable GameState where
  render :: GameState -> Picture
  render gstate = renderScreen (gameScreen gstate) gstate

-- Renders a spaceship (player/enemy)
instance Renderable Spaceship where
  render :: Spaceship -> Picture
  render p = translate x y (rotate angle (pictures [playerShape, flame]))
              where
                  c = if shipType p == PlayerShip then green else red
                  playerShape = Color c (Polygon [(-15,-10), (-15,10), (15,0)])
                  (x,y) = shipPosition p
                  angle = shipAngle p
                  flame = case shipState p of{Idle -> blank;
                                              Propeling -> Translate (-15) 0 (Scale 1 1 (Color yellow (Polygon [(4,7),(-4,0),(4, -7)])))}

-- Renders an asteroid                                              
instance Renderable Asteroid where
  render :: Asteroid -> Picture
  render a = translate x y (scale size size (pictures [asteroidShape]))
              where
                asteroidShape = Color cyan (Polygon [(0.0, 140.0),(140.0, 70.0), (140.0, -70.0),(0.0,-140.0), (-140.0, -70.0),(-140.0, 70.0)])
                (x,y)         = asPosition a
                size          = asSize a                                

-- Renders a missile                
instance Renderable Missile where
  render :: Missile -> Picture
  render m = translate (x+6.0) y (pictures [missileShape])
              where
                missileShape = Color yellow (Circle 3)
                (x,y)        = mslPosition m

-- Renders an animation of an asteroid breaking                
instance Renderable Animation where
  render :: Animation -> Picture
  render a = pictures pieces
              where
                missileShape =Color violet (Polygon [(-8.0,-2.0), (-8.0,2.0), (8.0,0.0)])
                (x,y)        = anPosition a
                pieces       = map ( \k -> translate (x+k) (y-k) missileShape ) (take (anTimeout a) [(-100),100,(-50),50,(-20),20,200,(-200),70,(-70)])

--------------------
-- Type class for renderable screens
--------------------
class RenderableScreen a where
  renderScreen :: a -> GameState -> Picture

instance RenderableScreen GameScreen where
  renderScreen :: GameScreen -> GameState -> Picture

  -- Renders the first screen with game logo and player name input
  renderScreen MainScreen gstate = translate (-280) 0 $ pictures[color white (text "Asteroids"), getHighscoreText gstate]

  -- Pause screen - displays the score and a message that the game is paused
  renderScreen PauseScreen gstate = translate (-230) 20 $ pictures [color white (text "Paused"), getStatsText gstate]

  -- End screen - displays "game over", players name and final score
  renderScreen EndScreen gstate = translate (-380) 20 $ pictures [color white (text "Game over!"), getHighscoreText gstate]

  -- Renders the game itself - player, enemies, asteroids, missiles and score
  renderScreen PlayScreen gState = translate (-640) (-360) $ pictures ( [mPlayer] ++ mEnemies ++ mAsteroids ++ mMissiles ++ mAnimations++ [mScore])
                              where
                                mPlayer       = render (player(gameObjects gState))
                                listEnemies   = enemies (gameObjects gState)
                                mEnemies      = map (render . enmShip) listEnemies
                                listAsteroids = asteroids (gameObjects gState)
                                mAsteroids    = map render listAsteroids
                                listMissiles  = missiles (gameObjects gState)
                                mMissiles     = map render listMissiles
                                listAnimations= animations (gameObjects gState)
                                mAnimations   = map render listAnimations
                                mScore        = getStatsTextGame gState

--------------------
-- Converts current game score into a picture
--------------------
getStatsText :: GameState -> Picture
getStatsText gstate = translate 0 (-120) $ color white (text ("Score:" ++ show (gameScore gstate)))

--------------------
-- Converts current game stats (name and score) into a picture
--------------------
getStatsTextGame :: GameState -> Picture
getStatsTextGame gstate = translate (20) (80) (scale 0.5 0.5 (getStatsText gstate))

--------------------
-- Converts current game highscore into a picutre
--------------------
getHighscoreText :: GameState -> Picture
getHighscoreText gstate = scale 0.5 0.5 $ translate 0 (-120) $ color white (text ("Highscore:" ++ show (gameHighscore gstate)))